play-json/shared/src/main/scala-3/play/api/libs/json/JsMacroImpl.scala

```diff
@@ -0,0 +1,146 @@
+          val disName = dis.drop(dis.lastIndexOf('.') + 1)
+          readCasesL[A, m.MirroredElemLabels, m.MirroredElemTypes](vjs, disName)
+        }
+      }
+    }
+  }
+
+  inline def writeCases[A: OWrites](x: A)(using m: Mirror.SumOf[A]): JsObject =
+    writeCasesL[A, m.MirroredElemLabels, m.MirroredElemTypes](x, m.ordinal(x))(0)
+
+  inline def readElemsL[A: Reads, L <: Tuple, T <: Tuple](obj: JsObject, elems: Array[Any])(n: Int)(using m: Mirror.ProductOf[A]): JsResult[A] =
+    inline (erasedValue[L], erasedValue[T]) match {
+      case _: (EmptyTuple, EmptyTuple) => JsSuccess(m.fromProduct(new ArrayProduct(elems)))
+      case _: (l *: ls, t *: ts)       => readElems1[A, l, t](obj) match {
+        case e @ JsError(_)  => e
+        case JsSuccess(x, _) =>
+          elems(n) = x
+          readElemsL[A, ls, ts](obj, elems)(n + 1)
+      }
+    }
+
+  inline def writeElemsL[A: OWrites, L <: Tuple, T <: Tuple](x: A)(n: Int, obj: JsObject): JsObject =
+    inline (erasedValue[L], erasedValue[T]) match {
+      case _: (EmptyTuple, EmptyTuple) => obj
+      case _: (l *: ls, t *: ts)       => writeElemsL[A, ls, ts](x)(n + 1, obj ++ writeElems1[A, l, t](x, n))
+    }
+
+  inline def readCasesL[A: Reads, L <: Tuple, T <: Tuple](js: JsValue, name: String): JsResult[A] =
+    inline (erasedValue[L], erasedValue[T]) match {
+      case _: (l *: ls, t *: ts)       =>
+        if (name == typeName[l]) summonReads[t & A].reads(js)
+        else readCasesL[A, ls, ts](js, name)
+      case _: (EmptyTuple, EmptyTuple) => JsError("error.invalid")
+    }
+
+  inline def writeCasesL[A: OWrites, L <: Tuple, T <: Tuple](x: A, ord: Int)(n: Int): JsObject =
+    inline (erasedValue[L], erasedValue[T]) match {
+      case _: (l *: ls, t *: ts)       =>
+        if (ord == n) {
+          val xjs = summonWrites[t].writes(x.asInstanceOf[t])
+          def jso = xjs match {
+            case xo @ JsObject(_) => xo
+            case jsv              => JsObject(Seq("_value" -> jsv))
+          }
+          JsObject(Map(config.discriminator -> JsString(typeName[l]))) ++ jso
+        } else writeCasesL[A, ls, ts](x, ord)(n + 1)
+      case _: (EmptyTuple, EmptyTuple) => throw new MatchError(x)
+    }
+
+  inline def readElems1[A: Reads, L, T](obj: JsObject): JsResult[T] = {
+    val reader = inline erasedValue[T] match {
+      case _: Option[a] => config.optionHandlers.readHandler(path[L])(summonReads[a]).asInstanceOf[Reads[T]]
+      case _            => path[L].read(summonReads[T])
+    }
+    reader.reads(obj)
+  }
+
+  inline def writeElems1[A: OWrites, L, T](x: A, n: Int): JsObject = {
+    val writer = inline erasedValue[T] match {
+      case _: Option[a] => config.optionHandlers.writeHandler(path[L])(summonWrites[a]).asInstanceOf[OWrites[T]]
+      case _            => path[L].write(summonWrites[T])
+    }
+    writer.writes(x.asInstanceOf[Product].productElement(n).asInstanceOf[T])
+  }
```

Comment:
As I wrote above to avoid allocating JsObject for every field...
```suggestion
  inline def writeElems1[A: OWrites, L, T](x: A, n: Int): collection.Map[String, JsValue] = {
    inline erasedValue[T] match {
      case _: Option[a] => {
        val writer = config.optionHandlers.writeHandler(path[L])(summonWrites[a]).asInstanceOf[OWrites[T]]
        writer.writes(x.asInstanceOf[Product].productElement(n).asInstanceOf[T]).underlying
      }
      case _            => Map.apply((config.naming(summonLabel[L]), summonWrites[T].writes(x.asInstanceOf[Product].productElement(n).asInstanceOf[T])))
    }
  }
```